File: Practica1/C++/Counting_Sort.cpp
#include <bits/stdc++.h>
#include "../GeneradorAleatorio/generator.h"
using namespace std;




void crear(int *&A,int n)
{
	A=new int[n];
}
void llenar(int *A,int n)
{
	for(int i=0;i<n;i++)
	{
		*(A+i)= rand()%1000;
	}
}
void imprimir(int *A,int n)
{
	for(int i=0;i<n;i++)
	{
		cout<<A[i]<<" ";
	}
	cout<<endl;
}
/////////////////////COUNTING SORT////////////////////////////
void countingSort(int *A,int tam)
{
	int max=*(A);
	int min=*(A);

	for(int i=0;i<tam;i++)
	{
		max=(*(A+i)>max)?*(A+i):max;
		min=(*(A+i)<min)? *(A+i):min;
	}

	int range=max - min +1;
	int *B=new int[range];

	for(int j=0;j<range;j++)
	{
		*(B+j)=0;
	}

	for(int i=0;i<tam;i++){
		B[A[i]-min]++;
//		*(B+(*(A+i)-min))++;
	}
	int p=0;
	for(int i=min;i<=max;i++)
	{
		for(int j=0;j<*(B+(i-min));j++)
		{
			*(A+(p++))=i;
		}
	}


}

int main()
{
	int n=10000;
	double t0;
	double time1;
	int k=0;
	ofstream fs("Tiempos/time_Counting.txt");
	srand(time(NULL));
	for(int i=n;i<=600000;i=i+60000){
		int *A;
		crear(A,i);
		read(A,i,k);
		t0=clock();
		countingSort(A,i);
		t0=clock()-t0;
		time1= ((double)t0)/CLOCKS_PER_SEC;
		k++;
		fs<<time1<<endl;
	}

	return 0;
}

File: Practica1/python/Counting_Sort.py
def countingSort(data):
    counts = [0 for i in range(max(data)+1)]

    for el in data:
        counts[el] += 1 

    for index in range(1, len(counts)):
        counts[index] = counts[index-1] + counts[index]

    L = [0 for loop in range(len(data))]
    for el in data:
        index = counts[el] - 1
        L[index] = el
        counts[el] -= 1 

	return L

print("Counting Loaded")
"""data = [4, 2, 2, 8, 3, 3, 1]
print("Array antes del cambio")
print(data)
countingSort(data)
print("Aplicando Counting Sort ")
print(data)"""
